Framework/Core/src/TableTreeHelpers.cxx

```diff
@@ -83,312 +95,259 @@ class ColumnIterator
+      mCurrent = new bool[listSize];
+      mLast = mCurrent + listSize;
+      accessChunk(0);
+    } else {
+      accessChunk(0);
+    }
   }
 
-  // create new branch of given data type
-  switch (mElementType) {
-    case arrow::Type::type::BOOL:
-      mLeaflistString += "/O";
-      break;
-    case arrow::Type::type::UINT8:
-      mLeaflistString += "/b";
-      break;
-    case arrow::Type::type::UINT16:
-      mLeaflistString += "/s";
-      break;
-    case arrow::Type::type::UINT32:
-      mLeaflistString += "/i";
-      break;
-    case arrow::Type::type::UINT64:
-      mLeaflistString += "/l";
-      break;
-    case arrow::Type::type::INT8:
-      mLeaflistString += "/B";
-      break;
-    case arrow::Type::type::INT16:
-      mLeaflistString += "/S";
-      break;
-    case arrow::Type::type::INT32:
-      mLeaflistString += "/I";
-      break;
-    case arrow::Type::type::INT64:
-      mLeaflistString += "/L";
-      break;
-    case arrow::Type::type::FLOAT:
-      mLeaflistString += "/F";
-      break;
-    case arrow::Type::type::DOUBLE:
-      mLeaflistString += "/D";
-      break;
-    default:
-      LOGP(FATAL, "Type {} not handled!", mElementType);
-      break;
+ private:
+  std::string mLeaflist;
+
+  void resetBuffer() override
+  {
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      if (O2_BUILTIN_UNLIKELY((*mCurrentPos - mFirstIndex) * listSize >= getCurrentArray()->length())) {
+        nextChunk();
+      }
+    } else {
+      if (O2_BUILTIN_UNLIKELY((mCurrent + (*mCurrentPos - mFirstIndex) * listSize) >= mLast)) {
+        nextChunk();
+      }
+    }
+    accessChunk(*mCurrentPos);
+    mBranch->SetAddress((void*)(mCurrent + (*mCurrentPos - mFirstIndex) * listSize));
   }
 
-  mBranchPtr = tree->Branch(mBranchName.c_str(), mBranchBuffer, mLeaflistString.c_str());
-  return mBranchPtr != nullptr;
-}
+  auto getCurrentArray() const
+  {
+    if (listSize > 1) {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(std::static_pointer_cast<arrow::FixedSizeListArray>(mColumn->chunk(mCurrentChunk))->values());
+    } else {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(mColumn->chunk(mCurrentChunk));
+    }
+  }
 
-bool BranchIterator::initDataBuffer(Int_t ib)
-{
+  void nextChunk() override
+  {
+    mFirstIndex += getCurrentArray()->length();
+    ++mCurrentChunk;
+  }
 
-  auto chunkToUse = mChunks.at(ib);
-  if (mFieldType == arrow::Type::type::FIXED_SIZE_LIST) {
-    chunkToUse = std::dynamic_pointer_cast<arrow::FixedSizeListArray>(chunkToUse)->values();
+  void accessChunk(int at)
+  {
+    auto array = getCurrentArray();
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      for (auto i = 0; i < listSize; ++i) {
+        mCurrent[i] = (bool)array->Value((at - mFirstIndex) * listSize + i);
+      }
+    } else {
+      mCurrent = (std::remove_pointer_t<T>*)array->raw_values();
```

Comment:
This is the only place where you really need T* with the current code, but even in that case you are basically using it just to get the size of a single element, which again you could do based on the field.